FindAllShortestPaths.cpp

```diff
@@ -0,0 +1,198 @@
+#include <algorithm>
+#include <iostream>
+#include <vector>
+#include <limits>
+#include <stack>
+#include <list>
+
+typedef std::vector<std::list<std::pair<int, int>>> adjacency_list; // std::pair<vertex, weight to that edge>
+
+// this function deletes the vertices that have more length than `ShortestDistance`
+void DeleteNotShortestPaths(std::vector<std::vector<std::pair<int, int>>>& Paths, int vertex, int ShortestDistance)
+{
+    // if vector is not empty and to your vertex path is longer than `ShortestDistance`, them we clear all that vector and start to fill it with `ShortestDistance`
+    if (!Paths[vertex].empty() && Paths[vertex].begin()->second > ShortestDistance)
+    {
+        Paths[vertex].clear();
+    }
+}
+
+// this Dijkstra is modified an dwill retunr matrix that has information about all shortest paths
+std::vector<std::vector<std::pair<int, int>>> ModifiedDijkstra(adjacency_list Graph, int start)
+{
+    std::vector<int> distance(Graph.size(), std::numeric_limits<int>::max());
+    std::vector<char> used(Graph.size(), 'f'); // never use vector of bools
+
+    // std::pair<vertex, distance from start to vertex>. `Parents[vertex]` shows from which verticies we can get to that vertex
+    std::vector<std::vector<std::pair<int, int>>> Parents(Graph.size());
+
+    distance[start] = 0;
+
+    for (int i = 0; i < Graph.size(); ++i)
+    {
+        int vertex = -1;
+
+        for (int j = 0; j < Graph.size(); ++j)
+        {
+            if (used[j] == 'f' && (vertex == -1 || distance[j] < distance[vertex]))
+            {
+                vertex = j;
+            }
+        }
+
+        if (distance[vertex] == std::numeric_limits<int>::max())
```

Comment:
Is there a guarantee that at this point `vertex` variable's value will not be `-1`?

Reply:
yes, there is a guarantee. Vertex == -1 || distance [j] <distance [vertex] and if we have vertex == -1 then the value j will be assigned to it